##%######################################################%##
#                                                          #
####            here we assign contig to sex            ####
####       chromosomes, investigate SNP density          ####
#                                                          #
##%######################################################%##

setwd("/Users/jean/Library/CloudStorage/OneDrive-UniversiteÃÅdePoitiers/recherche/sex_determinism/symchrosex/M_gene/scripts_SNPs")
source("WZ_functions.R")

# We compare nrec to simulated data in order to assign contigs to WZ chromosomes ---------------

# we import the results from stage 03
probsfams1 <- fread("../tables/fam1/countsAndPosteriors.txt")
probsfams2 <- fread("../tables/fam2/countsAndPosteriors.txt")

probs = merge(probsfams1, probsfams2, by = c("contig","length"),suffixes = c(".fam1",".fam2"))
probs[,c("p","nrec") := .(p.fam1 * p.fam2, nrec.fam1 + nrec.fam2)]

contigLengths = fread("../contigLengths.txt")
probs = probs[match(contigLengths$contig, contig),]

writeT(probs, "../tables/countsAndPosteriors.txt")

# we add a logical column indicating whether a contigs has informative SNPs in both families
# we do it indirectly by assessing whether the max posterior is higher than 1/11 (the prior)
probs[, two := pmin(daughters.r1.fam1 + daughters.r2.fam1, 
                    daughters.r1.fam2 + daughters.r2.fam2,
                    sons.r1.fam1 + sons.r2.fam1,
                    sons.r1.fam2 + sons.r2.fam2) > 0L]

# we do the same for contigs having SNPs in sons because only these contigs were used in simulations
probs[, one :=   sons.r1.fam1 + sons.r2.fam1 > 0L | sons.r1.fam2 + sons.r2.fam2 > 0L]

# we import all the values of nrec obtained by simulations
simu <- readRDS("../simu/nrec_1000repl.RDS")
simu <- data.table(contig = rep(which(probs$one), 1000), nrec = simu / 10^7)

# we compute the nrec 1/1000 quantile for each contig
quantiles <- simu[, .(q001 = quantile(nrec, 1 / 1000)), by = contig]

# we add this value to the table of actual contigs to assign to sex chromosomes
probs[one == T, q001 := quantiles$q001]

# a contig is assigned to the WZ chromosomes if its nrec is lower than the quantile
probs[, WZ := nrec < q001]

# we plot the distribution of nrec ------------------------------
# the lengths of contig categories to show in the plot
lengths <- probs[two == T, round(c(sum(length), sum(length[WZ])) / 10^6, 1)]

# colors of the polygons (observed, assigned to sex chromosomes and simulated contigs)
colors <- c(grey(0.3), "cadetblue", rgb(0.9, 0.4, 0.4, 0.6))

pdf("../figure1.pdf", 8, 7)

# we prepare the plot area
nSib = 96L
d <- probs[two == T, density(nrec / nSib * 100, from = 0, to = 50, bw = 0.1)]
probs[two == T, plot(d, type = "n", bty = "n", ylim = range(d$y), las = 1, 
                     main = "", xlab = "Inferred distance to the M gene (cM)")]

# we plot the density for observed data (all contigs)
closedPolygon(d$x, d$y, border = NA, col = colors[1])

# to plot the distribution for assigned contigs, we need a scale factor for the Y axis
Yscale <- probs[two == T, mean(WZ)]
probs[two == T & WZ == T, densityPolygon(nrec / nSib * 100, from = 0, to = 50, bw = 0.1, 
                                         Yscale = Yscale, border = NA, col = colors[2])]

# we plot the distribution for simulated data
simu[contig %in% which(probs$two), 
     densityPolygon(nrec / nSib * 100, from = 0, to = 50, bw = 0.1, border = NA, col = colors[3])]

legend(
  "topleft",
  border = NA,
  legend = c("Simulated data", stri_c(c("All analyzed contigs", "Contigs linked to the M gene"), " (~", lengths, " Mbp)")),
  fill = colors[c(3, 1, 2)], box.col = "white", bg = "white", y.intersp = 1.3, cex = 0.8, text.col = grey(0.3)
)

dev.off()

# we reclaim some RAM
rm(simu)

# we extrapolate chromosome length, given that not all contigs have SNPs in both families.


probs[, .(sum(length[two]), sum(length[which(WZ)]) / sum(length[two]) * sum(length))]


# investigating the divergence of m and M alleles via male heterozygous SNP density ---------------------------

# we import SNP calling files generated by angsd
beagle =fread("0-F0/fam1_2_mother_father_remapped_gatk.beagle", col.name = c("contig","allele1","allele2","fam_1_mo_1","fam_1_mo_2","fam_1_mo_3","fam_1_fa_1","fam_1_fa_2","fam_1_fa_3","fam_2_mo_1","fam_2_mo_2","fam_2_mo_3","fam_2_fa_1","fam_2_fa_2","fam_2_fa_3"))

counts  = fread("0-F0/fam1_2_mother_father_remapped_gatk.counts", col.names = c("fam_1_mo_A","fam_1_mo_C","fam_1_mo_G","fam_1_mo_T","fam_1_fa_A","fam_1_fa_C","fam_1_fa_G","fam_1_fa_T","fam_2_mo_A","fam_2_mo_C","fam_2_mo_G","fam_2_mo_T","fam_2_fa_A","fam_2_fa_C","fam_2_fa_G","fam_2_fa_T","va"))

pos= fread("0-F0/fam1_2_mother_father_remapped_gatk.pos")


proto_vcf = data.table(pos, beagle, counts)

proto_vcf = proto_vcf[,.(chr,pos,allele1,allele2,fam_1_fa_2,fam_2_fa_2,fam_1_fa_A,fam_1_fa_C,fam_1_fa_G,fam_1_fa_T,fam_2_fa_A,fam_2_fa_C,fam_2_fa_G,fam_2_fa_T)]

# computes sequencing depth on fathers as we discard SNPs with insufficient/excessive depths
proto_vcf = proto_vcf[,totDepth_fam_1_fa := fam_1_fa_A+fam_1_fa_C+fam_1_fa_G+fam_1_fa_T]
proto_vcf = proto_vcf[,totDepth_fam_2_fa := fam_2_fa_A+fam_2_fa_C+fam_2_fa_G+fam_2_fa_T]

depthQuantiles = proto_vcf[, .(fam1 = quantile(totDepth_fam_1_fa, 0.95), fam2 = quantile(totDepth_fam_2_fa, 0.95))]

proto_vcf[, hetFa1 := fam_1_fa_2 >= 0.99 & totDepth_fam_1_fa < depthQuantiles$fam1 & totDepth_fam_1_fa >= 5L]
proto_vcf[, hetFa2 := fam_2_fa_2 >= 0.99 & totDepth_fam_2_fa < depthQuantiles$fam2 & totDepth_fam_2_fa >= 5L]

SNPperContig = proto_vcf[, .(oneFather = sum(hetFa1 | hetFa2), bothFathers = sum(hetFa1 & hetFa2)), by = chr]

writeT(depthQuantiles, "tables/depthQuantileFathers.txt")

setnames(SNPperContig, 1, "CHROM")

probs = fread("../tables/countsAndPosteriors.txt")

# to estimate heterozygous SNP density per kb, we need to obtain the number of
# positions in the range of acceptable sequencing depths, for each contig. 
# We use the script below
system("Rscript nbValidPositions.R 10")

# we import the results
validPositions <- fread("../tables/nbValidPositionsFathers.txt")

# we merge both tables
SNPperContig <- merge(SNPperContig, validPositions, by = "CHROM", all = T)
SNPperContig[is.na(oneFather), oneFather := 0L]
SNPperContig[is.na(bothFathers), bothFathers := 0L]

# we add heterozygous SNP density columns to our general result table
probs[, c("het", "nPos") := 
        SNPperContig[match(contig, CHROM), .(oneFather * 1000 / nPos.V1, nPos.V1)]]

# we select contigs for which the max posterior probability of f is
# higher than 0.3 in all pools
used <- probs[, pmin(daughters.pM.fam1, daughters.pM.fam2, sons.pM.fam1, sons.pM.fam2) > 0.3]


# for the figure, we will compute heterozygous SNP density for different classes
# of nrec (on sex chromosomes) which we delineate by nrec deciles
nrecDeciles <- quantile(probs[WZ == T & used, nrec], probs = seq(0, 1, length.out = 11))

# we add a column to denote the class of nrec
probs[WZ == T & used, nrecClass := .bincode(nrec, nrecDeciles, include.lowest = T)]
meanNrecs <- probs[WZ == T & used, mean(nrec), by = .(class = nrecClass)]
probs[WZ == T & used, meanNrec := meanNrecs[match(nrecClass, class), V1]]

probs[, cM := nrec / nSib * 100]


pdf("../figure6.pdf", 8, 6)

layout(rbind(1:2), widths = 2:1)
par(mai = c(0.8, 0.8, 0.8, 0.4), cex.main = 0.9, oma = c(0, 0, 0, 0))

# we draw the scatter plot for each sex-chromosome contig
with(probs[WZ == T & used == T], ggBackground(
    cM, het, ylab = "Heterozygous SNP density in males (SNPs / kbp)", 
    xlab = "Inferred distance to the M gene (cM)", ylim = c(0, max(het)),
    main = stri_c("Contigs linked to the M gene (n=", sum(WZ), ")")
    ))

with(probs[WZ == T & used == T], points(cM, het, pch = 16, col = colors[2]))

# we add points (diamonds) for the nrec classes
distri <- probs[WZ == T & used == T, weighted.mean(het, nPos), by = meanNrec / nSib * 100]
distri <- probs[WZ == T & used == T, median(het), by = meanNrec / nSib * 100]
points(distri, pch = 23, cex = 1.5, bg = NA)

# we prepare and draw the violin plot (but we don't use ggplot here)

# to properly "align" with the previous plot,
# we get the max Y value from that plot
maxHet <- probs[WZ == T & used == T, max(het)]

# we compute the density
d <- probs[WZ == F & used == T & nrec > 0, 
           density(het, from = 0, to = maxHet, weights = nPos / sum(nPos))]

# we prepare the XY coordinates of the violon (polygon)
X <- c(d$y, -rev(d$y))
Y <- c(d$x, rev(d$x))
par(mai = c(0.8, 0, 0.8, 0.4))

# we plot the density
ggBackground(
    range(X), range(Y), ylim = c(0, maxHet), axes = F, xlab = "Density", 
    ylab = "", xgrid = F, main = stri_c("Other contigs (n=", d$n, ")")
    )

polygon(X, Y, border = NA, col = colors[1])

# we add a boxplot
with(probs[WZ == F & used == T & nrec > 0, ], boxplot(
    het, outline = F, add = T, at = 0, axes = F, col = "white", 
    boxwex = diff(range(X)) / 2, notch = T, lty = 1, staplewex = 0))

dev.off()


# we compute the correlation between heterozygous SNP density and distance to the SDR
probs[WZ == T & used == T, cor(het, cM)]

# we test the correlation
probs[WZ == T & used == T, cor.test(het, cM, method = "spearman", alternative = "less")]


# we report medians of contigs assigned to the SDR and other sex-chromosome contigs
probs[WZ == T & used, median(het), by = p > 0.5]

# we compare these medians with a Mann-Whitney test
probs[WZ == T & used, wilcox.test(het[p > 0.5], het[p <= 0.5], alternative = "greater")]

# and we compare heterozygous SNP density between contigs assigned to sex chromosomes and others
probs[, wilcox.test(het[used & WZ == T], het[used & WZ == F & nrec > 0], alternative = "two.sided")]

writeT(probs, "../tables/countsAndPosteriors.txt")
